benchbuild/experiment.py

```diff
@@ -134,18 +135,62 @@ def actions_for_project(self, project):
         """
 
     def actions(self):
+        """
+        Common setup required to run this experiment on all projects.
+        """
         actions = []
 
         for project in self.projects:
-            p = self.projects[project](self)
-            actions.append(Clean(p))
-            actions.append(MakeBuildDir(p))
-            project_actions = self.actions_for_project(p)
-            actions.append(Containerize(obj=p, actions=project_actions))
+            prj_cls = self.projects[project]
+
+            prj_actions = []
+            for version in self.sample(prj_cls, prj_cls.versions()):
+                p = prj_cls(self, version=version)
+
+                atomic_actions = [
+                    Clean(p),
+                    MakeBuildDir(p),
+                    Echo(message="Selected {0} with version {1}".format(
+                        p.name, p.version)),
+                    Containerize(obj=p, actions=self.actions_for_project(p))
+                ]
+                prj_actions.append(RequireAll(actions=atomic_actions))
+            actions.extend(prj_actions)
 
         actions.append(CleanExtra(self))
         return actions
 
+    def sample(self, prj_cls, versions):
+        """
+        Sample all avilable versions.
+
+        The default sampling returns the latest version only.
+        You can override this behavior in custom experiments.
+
+        Args:
+            prj_cls (class of benchbuild.project.Project):
+                The project class we sample versions for.
+            versions (list of str):
+                An ordered list of versions for the given prj_cls.
+                You can assume that the versions in this list are
+                valid version numbers for prj_cls.
+
+        Returns:
+            A list of version strings.
+
+        Assumptions:
+            versions:
+                * semantically ordered, descending.
+                * for version in versions:
+                    version in prj_cls.versions()
+        """
+        versions = prj_cls.versions()
```

Comment:
Found a bug here.
Passing versions into this function does not affect the output because we overwrite it with the versions the project provides.
How do you want to fix this? Default versions to `None` and only use prj_cls.versions() if it is `None` otherwise use the provided version list?

Reply:
>  Default versions to None and only use prj_cls.versions() if it is None otherwise use the provided version list?

Yep, let's do that.